src/app/qgsgeometryvalidationservice.cpp

```diff
@@ -0,0 +1,454 @@
+          break;
+        }
+      }
+    }
+  }
+
+  layer->triggerRepaint();
+
+  emit topologyErrorUpdated( layer, error );
+}
+
+void QgsGeometryValidationService::onLayersAdded( const QList<QgsMapLayer *> &layers )
+{
+  for ( QgsMapLayer *layer : layers )
+  {
+    QgsVectorLayer *vectorLayer = qobject_cast<QgsVectorLayer *>( layer );
+    if ( vectorLayer )
+    {
+      connect( vectorLayer->geometryOptions(), &QgsGeometryOptions::checkConfigurationChanged, this, [this, vectorLayer]()
+      {
+        enableLayerChecks( vectorLayer );
+      }, Qt::UniqueConnection );
+
+      connect( vectorLayer, &QgsVectorLayer::destroyed, this, [vectorLayer, this]()
+      {
+        cleanupLayerChecks( vectorLayer );
+        mLayerChecks.remove( vectorLayer );
+      } );
+
+      enableLayerChecks( vectorLayer );
+    }
+  }
+}
+
+void QgsGeometryValidationService::onFeatureAdded( QgsVectorLayer *layer, QgsFeatureId fid )
+{
+  if ( !mLayerChecks[layer].topologyChecks.empty() )
+  {
+    invalidateTopologyChecks( layer );
+  }
+  processFeature( layer, fid );
+}
+
+void QgsGeometryValidationService::onGeometryChanged( QgsVectorLayer *layer, QgsFeatureId fid, const QgsGeometry &geometry )
+{
+  Q_UNUSED( geometry )
+  // It would be nice to use the geometry here for the tests.
+  // But:
+  //  1. other codepaths to the checks also have no geometry (feature added / feature deleted)
+  //  2. and looking it up from the edit buffer (in memory) is really fast.
+  // so in short: it's still a good idea, but not as important as on first thought.
+
+  if ( !mLayerChecks[layer].topologyChecks.empty() )
+  {
+    invalidateTopologyChecks( layer );
+  }
+
+  processFeature( layer, fid );
+}
+
+void QgsGeometryValidationService::onFeatureDeleted( QgsVectorLayer *layer, QgsFeatureId fid )
+{
+  if ( !mLayerChecks[layer].topologyChecks.empty() )
+  {
+    invalidateTopologyChecks( layer );
+  }
+
+  mLayerChecks[layer].singleFeatureCheckErrors.remove( fid );
+
+  // There should be no geometry errors on an inexistent feature, right?
+  emit geometryCheckCompleted( layer, fid, QList<std::shared_ptr<QgsSingleGeometryCheckError>>() );
+}
+
+void QgsGeometryValidationService::onBeforeCommitChanges( QgsVectorLayer *layer )
+{
+  if ( !mLayerChecks[layer].topologyChecks.empty() ) // TODO && topologyChecks not fulfilled
+  {
+    if ( !layer->allowCommit() )
+    {
+      showMessage( tr( "Running geometry validation checks before saving ..." ) );
+    }
+
+    mLayerChecks[layer].commitPending = true;
+
+    triggerTopologyChecks( layer );
+  }
+}
+
+void QgsGeometryValidationService::onEditingStopped( QgsVectorLayer *layer )
+{
+  cancelTopologyCheck( layer );
+  clearTopologyChecks( layer );
+}
+
+void QgsGeometryValidationService::showMessage( const QString &message )
+{
+  mMessageBar->popWidget( mMessageBarItem );
+  mMessageBarItem = QgsMessageBar::createMessage( tr( "Geometry Validation" ),  message );
+  mMessageBarItem->setDuration( 5 );
```

Comment:
Better to use the default duration here

Reply:
That was the intention (https://github.com/qgis/QGIS/blob/master/src/gui/qgsmessagebar.h#L88) is there a safer way to do this?